mixing/mixpool/observer.go

```diff
@@ -0,0 +1,503 @@
+// Copyright (c) 2025 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mixpool
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/wire"
+)
+
+const strikeLimit = 2
+
+const minPeers = 4 // Keep synced with mixclient.MinPeers
+
+type strikeSet struct {
+	set     map[wire.OutPoint]struct{}
+	strikes []uint64
+}
+
+func sortUniq(strikes []uint64) []uint64 {
+	sort.Slice(strikes, func(i, j int) bool {
+		return strikes[i] < strikes[j]
+	})
+	n := len(strikes)
+	strikes = strikes[:1]
+	last := strikes[0]
```

Comment:
This looks rather dangerous.  It creates the requirement for sortUniq to be called with a non-empty slice and there are no comments on the function calling out a panic condition.